python/hackathon_app/sheets.py

```diff
@@ -0,0 +1,113 @@
+from datetime import datetime
+
+from google.oauth2 import service_account  # type: ignore
+from googleapiclient import discovery  # type: ignore
+
+from typing import Sequence, List, Dict
+
+
+class Sheets:
+    """An API for manipulating the Google Sheet containing hackathon data."""
+
+    def __init__(self, spreadsheet_id):
+        scopes = [
+            "https://www.googleapis.com/auth/drive",
+            "https://www.googleapis.com/auth/drive.file",
+            "https://www.googleapis.com/auth/spreadsheets",
+        ]
+
+        credentials = service_account.Credentials.from_service_account_file(
+            "jaxdata-a93316beb3a0.json", scopes=scopes
+        )
+
+        service = discovery.build("sheets", "v4", credentials=credentials)
+        self.spreadsheet = service.spreadsheets().values()
+        self.id = spreadsheet_id
+
+    def get_hackathons(self):
+        """Get names of active hackathons."""
+        hackathons_table = self.spreadsheet.get(
+            spreadsheetId=self.id, range="hackathons!A1:end"
```

Comment:
Probably planned, but the `range` value should probably be declared in a constant somewhere then ref'd in these accessors. I'd possibly even consider a `Hackathon` object that supports the CRUD operations for it, and a `User` object, etc., where the `rangetag` is a property of the class. 

Create a `WhollySheet` base class:
```python
class WhollySheet:
  rangeid: str
  def items -> Sequence[SheetRow] # all matching values. maybe call it rows instead

class SheetRow: # or SheetItem
  def create(args) -> appended obj
  def copy() -> copied obj
  def update(selector, values) -> updated obj
  def delete(selector) -> true if deleted, possibly throw error if not
  def get(selector) -> obj
  def search() -> Sequence[obj] # or find() .. not sure if this is feasible, can wait
```
Something along those lines as a first cut. Could definitely be done with generics, too.

Then you have `Hackathon`, `Hackathons, `Users`, and `User` objects as descendants that you can use to manipulate the row type as a class.

```python
class Hackathon(SheetRow):
  # define all hackathon-specific attributes

class Hackathons(WhollySheet) : 
  rangeid = "hackathons!A1:end" # or we just care about the tab name, so it would be "hackathons" with "!A1:end" standard

class User(SheetRow):
  # define all user-specific attributes

class Users(WhollySheet):
  rangeid = "users!A1:end"
```

This should also make writing the tests for all relevant operations much easier as well. And it abstracts away our storage mechanism.

Reply:
I agree that this is a good approach (and we'd talked about even using `attrs` + `cattrs` to easily hydrate `User` and `Hackathon` objects).

However, I'd love to glue this in asap to the frontend. I think the top level contract (`get_hackathons() -> str` and `register_user(args) -> bool` is good enough to merge this into `python/hackathon_app` and then a separate PR to refactor the internal data/object handling and tests

Reply:
I agree with both. I'll get this in a state where we can merge it and then I'll submit a separate PR with the discussed improvements.